### PR TITLE
Extend label format to include index labels

### LINKED ISSUE #126: Incompatibility between ART API for DataGenerators and standard PyTorch data loaders
Existing standard PyTorch data loaders return ordinal encoded labels, as is usual practice for this framework. ART `DataGenerators` expect / assume one-hot encoded labels. For this reason, the standard data loaders can't be wrapped under the `PyTorchDataGenerator` class.

Changing `PyTorchDataGenerator` to accept ordinal labels would break the ART API and would require verifying the format of the labels in many places. An alternative solution is to convert the labels to one-hot encoding if necessary inside the `PyTorchDataGenerator`. This would require to add the total number of classes as parameter to the constructor of the wrapper. 

This issue & solution were proposed by Andras Rozsa.

### PR DESCRIPTION
**Pull request by beat-buesser**
This commit extends ART at `attack.generate`, `classifier.fit` and `classifier.loss_gradient` to accept now index labels and one-hot-encoded labels (as before). Internally ART continues to treat targets as one-hot-encoded. That way users of ART can use ART with the label format preferred by their machine learning framework.

Fixes #126 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

I have been able to run example script data_augmentation.py with `torchvision.datasets.MNIST`, `torch.utils.data.DataLoader` and `PyTorchDataGenerator`.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/projected_gradient_descent.py [hunk 1]
```diff
@@ -90,14 +90,16 @@ def generate(self, x, y=None, **kwargs):
 
         :param x: An array with the original inputs.
         :type x: `np.ndarray`
-        :param y: The labels for the data `x`. Only provide this parameter if you'd like to use true
-                  labels when crafting adversarial samples. Otherwise, model predictions are used as labels to avoid the
-                  "label leaking" effect (explained in this paper: https://arxiv.org/abs/1611.01236). Default is `None`.
-                  Labels should be one-hot-encoded.
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
```
Comment (ririnicolae) [inline art/attacks/projected_gradient_descent.py]: You could use backquotes for shapes in docstrings. Same goes wherever this docstring was copied.
```suggestion
        :param y: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)` or indices of shape
```

File: art/classifiers/catboost.py [hunk 2]
```diff
@@ -66,7 +66,8 @@ def fit(self, x, y, **kwargs):
 
         :param x: Training data.
         :type x: `np.ndarray`
-        :param y: Labels, one-vs-rest encoding.
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or inices of shape
```
Comment (ririnicolae) [inline art/classifiers/catboost.py]:
```suggestion
        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
```

File: art/attacks/hop_skip_jump.py [hunk 3]
```diff
@@ -87,13 +87,16 @@ def generate(self, x, y=None, **kwargs):
 
         :param x: An array with the original inputs to be attacked.
         :type x: `np.ndarray`
-        :param y: If `self.targeted` is true, then `y` represents the target labels.
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
```
Comment (ririnicolae) [inline art/attacks/hop_skip_jump.py]:
```suggestion
        :param y: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)` or indices of shape
```

File: art/attacks/saliency_map.py [hunk 4]
```diff
@@ -69,11 +70,14 @@ def generate(self, x, y=None, **kwargs):
 
         :param x: An array with the original inputs to be attacked.
         :type x: `np.ndarray`
-        :param y: Target values if the attack is targeted
+        :param y: Target values (class labels) one-hot-encoded of shape (nb_samples, nb_classes) or indices of shape
```
Comment (ririnicolae) [inline art/attacks/saliency_map.py]:
```suggestion
        :param y: Target values (class labels) one-hot-encoded of shape `(nb_samples, nb_classes)` or indices of shape
```

File: art/attacks/saliency_map.py [hunk 5]
```diff
@@ -69,11 +70,14 @@ def generate(self, x, y=None, **kwargs):
...
+                  (nb_samples,).
```
Comment (ririnicolae) [inline art/attacks/saliency_map.py]:
```suggestion
                  `(nb_samples,)`.
```

File: art/classifiers/scikitlearn.py [hunk 6]
```diff
@@ -770,28 +784,6 @@ def class_gradient(self, x, label=None, **kwargs):
 
         return gradients
 
-    def fit(self, x, y, **kwargs):
```
Comment (ririnicolae) [inline art/classifiers/scikitlearn.py]: Why were the `fit` functions removed and how is this related to the scope of this PR?

Comment (beat-buesser) [reply art/classifiers/scikitlearn.py]: We have looked at the removed `fit` functions during development for this PR to see if we need to update them with `check_and_transform_label_format`, and noticed that they can be made redundant and included into the `fit` function of the parent class `ScikitlearnClassifier` where we have `check_and_transform_label_format` as part of `_apply_preprocessing`.

**Comment by beat-buesser** [comment]
@ririnicolae @mathsinn Thank you for your reviews and great comments!
